File: solutions/c/high-scores/1/high_scores.c
#include "high_scores.h"


int32_t latest(const int32_t *scores, size_t len)
{
    return scores[len - 1];
}

int32_t personal_best(const int32_t *scores, size_t len)
{
    int32_t best = 0;
    for(size_t i = 0; i < len; i++)
    {
        if(scores[i] > best)
            best = scores[i];
    }
    return best;
}

size_t personal_top_three(const int32_t *scores, size_t len, int32_t *output)
{
    int32_t best = 0;
    size_t loc1, loc2;
    
    for(size_t i = 0; i < len; i++)
    {
        if(scores[i] > best)
        {
                best = scores[i];
                loc1 = i;
        }
    }
    output[0] = best;

    if(len == 1)
        return 1;

    best = 0;
    for(size_t i = 0; i < len; i++)
    {
        if(scores[i] > best)
        {
                if(i != loc1)
                {
                    best = scores[i];
                    loc2 = i;
                }
        }
    }
    output[1] = best;

    if(len == 2)
        return 2;

    best = 0;
    for(size_t i = 0; i < len; i++)
    {
        if(scores[i] > best)
        {
                if(i != loc1 && i != loc2)
                {
                    best = scores[i];
                }
        }
    }
    output[2] = best;

    return 3;
}

File: solutions/java/high-scores/1/src/main/java/HighScores.java
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;

class HighScores
{
    private List<Integer> scoresList;
    
    public HighScores(List<Integer> highScores)
    {
        this.scoresList = new ArrayList<>(highScores);
    }

    List<Integer> scores()
    {
        return List.copyOf(this.scoresList);
    }

    Integer latest()
    {
        return this.scoresList.get(this.scoresList.size() - 1);
    }

    Integer personalBest()
    {
        int max = 0;
        for (Integer i: this.scoresList)
        {
            if (i > max)
            {
                max = i;
            }
        }
        return max;
    }

    List<Integer> personalTopThree()
    {
        Integer a = 0, b = 0, c = 0;
        for (Integer i: this.scoresList)
        {
            if (i > a)
            {
                c = b;
                b = a;
                a = i;
            }
            else if (i > b)
            {
                c = b;
                b = i;
            }
            else if (i > c)
            {
                c = i;
            }
        }
        
        switch (this.scoresList.size())
        {
            case 1:
                return Arrays.asList(a);

            case 2:
                return Arrays.asList(a, b);

            default:
                return Arrays.asList(a, b, c);
        }
    }
}
